lib/src/mailmap.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Support for `.mailmap` files.
+
+use std::fmt::Debug;
+use std::io::{self, Read};
+
+use pollster::FutureExt;
+
+use crate::backend::Signature;
+use crate::commit::Commit;
+use crate::conflicts::{materialize_tree_value, MaterializedTreeValue};
+use crate::op_store::WorkspaceId;
+use crate::repo::Repo;
+use crate::repo_path::RepoPath;
+
+/// Models a `.mailmap` file, mapping email addresses and names to
+/// canonical ones.
+///
+/// The syntax and semantics are as described in
+/// [`gitmailmap(5)`](https://git-scm.com/docs/gitmailmap).
+///
+/// You can obtain the currently‐applicable [`Mailmap`] using
+/// [`get_current_mailmap`].
+///
+/// An empty [`Mailmap`] does not use any heap allocations, and an absent
+/// `.mailmap` file is semantically equivalent to an empty one, so there is
+/// usually no need to wrap this type in [`Option`].
+#[derive(Clone, Default)]
+pub struct Mailmap(gix_mailmap::Snapshot);
+
+impl Debug for Mailmap {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("Mailmap").field(&self.0.entries()).finish()
+    }
+}
+
+impl PartialEq for Mailmap {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.entries() == other.0.entries()
+    }
```

Comment:
nit: might be better to use pointer-based comparison, and add TODO comment.
```// TODO: Refactor tests to not need the Eq trait```

`.entries()` constructs `Vec<_>`.

Reply:
Would moving the tests over to `matches!` in another commit here work?

I’ve opened https://github.com/Byron/gitoxide/issues/1424 to ask for a better solution upstream; I might just send a PR so we can avoid the allocation here. Non‐semantic equality on immutable types makes me a bit nervous, personally.

Reply:
> Would moving the tests over to `matches!` in another commit here work?

I have draft patches to do that (moving to `assert_matches!` or `insta::assert_debug_snapshot`), so feel free to leave a TODO comment.

Reply:
Done!

Reply:
`gix-mailmap` 0.23.3 now has `Debug` and `Eq` implementations that don’t allocate, so I’ve switched to a `#[derive]` here. Adjusting the tests to not use equality comparison may still be a good idea, of course.